sim/index.js

```diff
@@ -0,0 +1,151 @@
+const RECOMMEND_CLICKTHROUGH_MAP = {
+  random: 0,
+  recommended: 1,
+};
+
+const config = {
+  // user
+  numOfUsers: DEFAULT.NUM_USERS,
+  userIdStart: DEFAULT.USER_ID_START,
+  // game
+  numOfGames: DEFAULT.NUM_GAMES,
+  gameIdStart: DEFAULT.GAME_ID_START,
+};
+
+// this is essentially the RTC for the simulation and is the source of truth
+// for all event times
+let CurrentSimTime = START_DATE_MS;
+
+class UserEvent {
+  constructor(type, userId, date, content = null) {
+    this.type = type;
+    this.userId = userId;
+    this.date = date;
+    this.content = content;
+  }
+  print() {
+    console.log(`-----------
+      type: ${this.type}
+      userId: ${this.userId}
+      date: ${this.date}
+      content: ${nodeUtil.inspect(this.content, { showHidden: false, depth: null })}
+    `);
+  }
+}
+
+const generateRandomUserEvent = (burstIntervalRate, burstQuantity) => {
+  let randomUserId;
+  let randomGameId;
+  let randomIsRecommendedGame;
+
+  const RECOMMEND_CLICKTHROUGH_RATIO_WEIGHT_TABLE =
+    util.generateExpandedWeightTable(RECOMMEND_CLICKTHROUGH_RATIO);
+
+  CurrentSimTime += burstIntervalRate;
+
+  // init the object with something just so we can change properties later
+  // and not have to recreate the object
+  const userEvent = new UserEvent(
+    'click',
+    randomUserId,
+    CurrentSimTime,
+    {
+      gameId: randomGameId,
+      isRecommendedGame: randomIsRecommendedGame,
+    }
+  );
+  for (let i = 0; i < burstQuantity; i += 1) {
+    randomUserId = util.getRandomNumberInclusive(config.userIdStart, config.numOfUsers);
+    randomGameId = util.getRandomNumberInclusive(config.gameIdStart, config.numOfGames);
+    randomIsRecommendedGame =
+      RECOMMEND_CLICKTHROUGH_MAP[
+        util.getRandomFieldValue(RECOMMEND_CLICKTHROUGH_RATIO_WEIGHT_TABLE)];
+
+    // reset the values of the object to be more memory efficient
+    userEvent.type = 'click';
+    userEvent.userId = randomUserId;
+    userEvent.date = CurrentSimTime;
+    userEvent.content = {
+      gameId: randomGameId,
+      isRecommendedGame: randomIsRecommendedGame,
+    };
+    // TODO: dispatch the event here
+    userEvent.print();
+  }
+};
+
+const startSimulation =
+  (burstRate = DEFAULT.BURST_RATE, burstQuantity = DEFAULT.BURST_QUANTITY) => {
+    setInterval(() => generateRandomUserEvent(burstRate, burstQuantity), burstRate);
+  };
+
+const displayUsage = () => {
+  console.log(`
+    Utility script used to simulate real-time User Events and is meant to be started and run in the background
+
+      [usage] node ${path.basename(__filename)} <EVENT_BURST_INTERVAL> <EVENT_BURST_QUANTITY>
+
+        EVENT_BURST_INTERVAL: interval in milliseconds to send a batch of events
+
+        EVENT_BURST_QUANTITY: amount of events to send at each burst interval
+    `);
+};
+
+// simple CLI
+// [usage] node index.js <EVENT_BURST_INTERVAL> <EVENT_BURST_QUANTITY>
+if (process.argv.length > 1) {
```

Comment:
Lam - this looks really good, overall not much to add. Would like to understand what process.argv is so maybe some additional comments on the CLI code for better understanding?

Reply:
Thanks for the feedback, @bnguyen2! process.argv is an array that contains the command-line arguments that get passed in when executing a script via the command-line. It allows you to do stuff like:
```myscript.js <arg1> <arg2>```

<arg1> and <arg2> in this case get pushed into that process.argv array so it can be utilized in the script. Think of it as the public "API" to the script.

You can also make changes to values into a script, but it's best not to have to tweak a script to do so. Using command-line arguments give you that flexibility without having to make code changes.

More details can be found here:
https://nodejs.org/docs/latest/api/process.html#process_process_argv